Implementation/64347 new project attribute calculated value

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/64347

# What are you trying to accomplish?

* Introducing calculated value as a new custom field format. Only available as a project attribute and only if the feature flag is active.
  * Note that the **feature flag** only controls the _creation_ of calculated values. With the feature flag disabled, you can still see calculated values that were created when the flag was active. You can also edit existing calculated values, regardless of the feature flag.
* Primerized view similar to the hierarchy. The view is rendered instead of the default custom field view, in the same way as #19182 established for work package custom fields.
* Basic formula validation
  * Checks for invalid characters
  * Checks mathematical or syntactic errors
* The formula is parsed and tokens like `cf_123` have the custom field ID (123 in this case) extracted into its own JSONB section `referenced_custom_fields`. But the ID is not validated in any way (e.g., does it belong to an actual CustomField the user has permission to access, etc.).

### Out of scope

* Displaying the new custom field in a useful way
* Actually calculating a value and storing it to the database

## Drive-by fixes

* There was an error in the breadcrumb and title display of the Hierarchy DetailsForm which I noticed because I introduced the same issue in the new Calculated Value Details Form. If you entered an empty name and saved the form, the validation would fail, which is good. But after page reload, the empty name of the in-memory model would be rendered in the breadcrumbs and title, which looked bad. Instead, it should display the persisted name. Fixed for both.
* Fixed some occurrences of custom field format labels without translation.

## Screenshots

![image](https://github.com/user-attachments/assets/3f6a78cf-b59d-4118-807c-0613d49db26b)

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#64347] add new format "calculated_value" to project attribute list

EinLama committed:
[#64347] add formula column (and index) to custom_fields

EinLama committed:
[#64347] primerize view for calculated value

EinLama committed:
[#64347] introduce formula field to custom field admin views

EinLama committed:
[#64347] translate custom field formats on index

EinLama committed:
[#64347] show correct active menu item

EinLama committed:
[#64347] controller checks feature flag

EinLama committed:
[#64347] check feature flag in default scope for custom fields

EinLama committed:
[#64347] remove non-existent module import

EinLama committed:
[#64347] write/read formula, prepare validation

EinLama committed:
[#64347] perform basic formula validation

EinLama committed:
[#64347] extract custom field IDs from formula

EinLama committed:
[#64347] remove unnecessary calculated values controller

With the menu rework, this is no longer necessary

EinLama committed:
[#64347] inject Primer view for calculated values into existing admin view

much like the hierarchy before

EinLama committed:
[#64347] get rid of inefficient regular expression

EinLama committed:
[#64347] move calculated value logic into its own module

EinLama committed:
[#64347] calculated value model specs

EinLama committed:
[#64347] remove falsy dot from token pattern

EinLama committed:
[#64347] error messages

EinLama committed:
[#64347] add missing translations for field formats

EinLama committed:
[#64347] project custom fields index spec

EinLama committed:
[#64347] refactor hidden calculated values w/o feature flag

EinLama committed:
[#64347] use proper translation for field format

EinLama committed:
[#64347] fix unused code and wrong edit path

EinLama committed:
[#64347] only use hidden fields for new records

EinLama committed:
[#64347] fix custom field Primer bugs

In the admin view, when you edit a custom fields name to be empty and
try to save that, the display would incorrectly assume the empty title
from the model, even if the model would be invalid with these changes.
Updated to fetch the name attribute from the database instead.

The bug only happens when you use a Primer component. So for now, this
is only true for Hierarchy and Calculated Value CFs.
Angular prevents this bug from happening since there is no page reload.

EinLama committed:
[#64347] details form spec, better model validation

EinLama committed:
[#64347] copyright

EinLama commented on app/components/projects/settings/project_custom_field_sections/custom_field_row_component.html.erb:
```diff
@@ -13,7 +13,7 @@
         title_container.with_column(pt: 1, mr: 2, data: { test_selector: "custom-field-type" }) do
           render(Primer::Beta::Text.new(font_size: :small, color: :subtle)) do
-            @project_custom_field.field_format.capitalize
+            t("label_#{@project_custom_field.field_format}")
```
Should use `custom_field_helper.label_for_custom_field_format` here instead

toy replied:
```suggestion
            helpers.label_for_custom_field_format(@project_custom_field.field_format)
```

EinLama commented on app/components/settings/project_custom_field_sections/custom_field_row_component.html.erb:
```diff
@@ -19,7 +19,7 @@
         content_container.with_column(mr: 2) do
           render(Primer::Beta::Text.new(font_size: :small, color: :subtle)) do
-            @project_custom_field.field_format.capitalize
+            t("label_#{@project_custom_field.field_format}")
```
Should use `custom_field_helper.label_for_custom_field_format` here instead

toy replied:
```suggestion
            helpers.label_for_custom_field_format(@project_custom_field.field_format)
```

toy commented on app/components/admin/custom_fields/calculated_values/details_form.rb:
Most forms live under `app/forms`

EinLama replied:
Thank you, I didn't notice that. It did feel off. Moved this to `app/forms`.

toy commented on app/components/admin/custom_fields/calculated_values/details_form.rb:
```diff
@@ -0,0 +1,78 @@
+        caption: I18n.t("custom_fields.instructions.admin_only")
+      )
+
+      details_form.submit(name: :submit, label: I18n.t(:button_save), scheme: :default)
```
This should have `scheme: :primary` and label should probably depending on `new_record?`
I pinged Marc, mockups are not correct

toy commented on app/components/admin/custom_fields/calculated_values/details_form.rb:
```diff
@@ -0,0 +1,78 @@
+        caption: I18n.t("custom_fields.instructions.formula")
+      )
+
+      details_form.check_box(
+        name: :is_for_all,
+        label: I18n.t("activerecord.attributes.custom_field.is_for_all"),
+        caption: I18n.t("custom_fields.instructions.is_for_all")
+      )
```
Should be `is_required`, as `is_for_all` is only for work package custom fields

EinLama replied:
Thank you for spotting this. Fixed!

toy commented on app/components/settings/project_custom_field_sections/show_component.rb:
```diff
@@ -38,6 +38,11 @@ def initialize(project_custom_field_section:, first_and_last: [])
+
+        unless OpenProject::FeatureDecisions.calculated_value_project_attribute_active?
+          @project_custom_fields = @project_custom_fields.where.not(field_format: "calculated_value")
+        end
```
Do you think already created fields should be hidden? I think it is better to prevent creation when flag is not on, but not hide already created custom fields

EinLama replied:
I was thinking about this, too... Now that you have also mentioned it, I put it up for discussion in the daily. We came to the conclusion that for feature flags, it is enough to prevent the creation of new calculated values as you suggested. For enterprise editions features - which calculated values will eventually turn out to be - the approach might be different. We have a very similar situation with Hierarchy custom fields, too. We will discuss it :) 

For now, limiting the feature flag to preventing the creation of new custom fields 👍 Thanks!

toy commented on app/components/admin/custom_fields/edit_form_header_component.html.erb:
```diff
@@ -29,7 +29,7 @@ See COPYRIGHT and LICENSE files for more details.
 <%=
   render(Primer::OpenProject::PageHeader.new(test_selector: "custom-fields--page-header")) do |header|
-    header.with_title { @custom_field.name }
+    header.with_title { @custom_field.attribute_in_database("name") || @custom_field.name }
```
I don't think there is a case when `@custom_field.attribute_in_database("name")` returns `nil`
Applies to all other places

EinLama committed:
[#64347] use existing helper for cf label

toy commented on app/components/admin/custom_fields/calculated_values/details_form.rb:
```diff
@@ -0,0 +1,78 @@
+        label: I18n.t("activerecord.attributes.project_custom_field.custom_field_section"),
+        required: true
+      ) do |li|
+        ProjectCustomFieldSection.all.collect.each do |cs|
```
Is `collect` changing anything?
```suggestion
        ProjectCustomFieldSection.all.each do |cs|
```

EinLama replied:
Thanks, replaced the entire call-chain with `find_each` -> `ProjectCustomFieldSection.find_each do |cs|`

toy commented on app/components/settings/project_custom_fields/edit_form_header_component.html.erb:
```diff
@@ -28,7 +28,7 @@ See COPYRIGHT and LICENSE files for more details.
 <%=
   render(Primer::OpenProject::PageHeader.new) do |header|
-    header.with_title { @custom_field.name }
+    header.with_title { @custom_field.attribute_in_database("name") || @custom_field.name }
```
```suggestion
    header.with_title { @custom_field.attribute_in_database("name") }
```

toy commented on lib/open_project/custom_field_format.rb:
```diff
@@ -72,8 +74,15 @@ def register(custom_field_format, _options = {})
+          formats
+        else
+          formats.except("calculated_value")
+        end
```
I'm wondering if it may be better to make it an attribute of the formatter, like `enabled` or `disabled`. Can be a proc, as you don't want to make it static after being set in the initialiser

toy commented on spec/contracts/project_custom_field_project_mappings/base_contract_spec.rb:
```diff
@@ -49,10 +49,12 @@
+      # Stub `all` to return an ActiveRecord::Relation, not an array, so that further query chaining
+      # (e.g., .where) works as expected.
+      allow(ProjectCustomField).to receive(:all).and_return(ProjectCustomField.where(id: project_custom_field.id))
     end
```
Is stubbing needed at all if record is created and not built?

EinLama replied:
Ah, of course not. The stub can be erased. Thanks!

toy commented on app/views/custom_fields/_form.html.erb:
```diff
@@ -76,6 +76,14 @@ See COPYRIGHT and LICENSE files for more details.
+    <span class="form--field-instructions">
+      <%= t("custom_fields.instructions.formula") %>
+    </span>
+  </div>
```
Why add it here if there is no case for it to be displayed?

EinLama replied:
Initially I wanted to consider other custom fields that might implement the formula at some point and keep the form partial consistent with the format dependent attributes. But as a formula is a special use case, this will most likely never happen. Removed this form element! ⌫

toy commented on db/migrate/20250613141234_add_formula_to_custom_fields.rb:
```diff
@@ -0,0 +1,37 @@
+
+class AddFormulaToCustomFields < ActiveRecord::Migration[8.0]
+  def change
+    add_column :custom_fields, :formula, :jsonb, null: true
+
+    add_index :custom_fields, :formula, using: :gin
```
Why do you prefer to store referenced cf ids in a structure alongside the formula string, and not have the column only for the formula and reference used cfs in a many-to-many table?

EinLama replied:
For the first iteration, I prefer this simple approach, as we do not know yet where exactly the implementation will take us. By storing it in a serialized format for now, we have all we need for the next steps and can start splitting up the work (i.e., the next work packages) quickly. If we have to extend the solution to use a many-to-many table, we can easily add that later.

toy commented on app/models/custom_field.rb:
```diff
@@ -72,6 +74,7 @@ def uniqueness_of_name_with_scope
 
   validate :validate_default_value
   validate :validate_regex
+  validate :validate_formula
```
I'm wondering if it would better be added from `CustomField::CalculatedValue` using concern helpers (or `self.included`)

EinLama replied:
Yes, good point. I also think that's more elegant. Made it so!

toy commented on app/models/custom_field.rb:
```diff
@@ -299,6 +302,10 @@ def field_format_hierarchy?
     field_format == "hierarchy"
   end
 
+  def field_format_calculated_value?
```
I think better just `calculated_value?` and it would be easier at some point to generate such methods for each available format instead of doing it manually

EinLama replied:
Yes that would for sure be an option in the future. For now, I wanted to keep the name consistent with `field_format_hierarchy?`. And since the Calculated Value might be the last custom field format requiring a frequent check for the format (due to UI differences), I figured I would not implement the generator just now. Two manual methods are fine.

toy commented on app/controllers/admin/settings/project_custom_fields_controller.rb:
```diff
@@ -208,6 +210,8 @@ def find_custom_field_projects_to_link
     def find_custom_field
       @custom_field = ProjectCustomField.find(params[:id])
+
+      render_404 if @custom_field.field_format_calculated_value? && !calculated_value_feature_flag?
```
Also here better to limit creation of new custom fields, but not limit editing existing

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+    # Dentaku will return nil if the formula is invalid.
+    # TODO WP-64348: add support for referenced custom fields by injecting them as variables,
+    #       e.g. Dentaku(formula_string, cf_123: CustomField.find(123).value)
+    errors.add(:formula, :invalid) unless Dentaku(formula_string)
```
I think it is better from the start to use `Dentaku::Calculator.new.evaluate(formula_string)` to not have odd errors with shared instance
Also maybe enough to just get to dependencies - `Dentaku::Calculator.new.dependencies(formula_string)`

EinLama replied:
Sure. Doesn't hurt to start using an instance right away. We will need it soon, anyways 👍

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+  # Returns a list of custom field IDs used in the formula.
+  # For a formula like `2 + cf_12 + cf_4` it returns `[12, 4]`.
+  def cf_ids_used_in_formula(formula_str = formula_string)
+    formula_str.scan(/cf_(\d+)/).flatten.map(&:to_i)
```
I suggest adding word boundaries to regexp (`/\bcf_(\d+)\b/`)

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+    # the custom field.
+    # Once this basic validation passes, the formula will be parsed and validated by Dentaku, which builds an AST
+    # and ensures that the formula is really valid. A welcome side effect of the basic validation done here is that
+    # it prevents built-in functions from being used in the formula, which we do not want to allow.
```
Why don't we want built-in functions?

EinLama replied:
Dentaku offers functions such as `IF` (like Excel), and so on. Since we won't be offering these functions in the formula editor for now, they should also not be allowed in the formula string as that might cause confusion.

We might add some of these later, but at the beginning, having a small subset of possibilities for this new feature seems like a good idea to me.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+    # and ensures that the formula is really valid. A welcome side effect of the basic validation done here is that
+    # it prevents built-in functions from being used in the formula, which we do not want to allow.
+    allowed_chars = %w[+ - / * ( )] + [" "]
+    split_pattern = /\s|(\+)|(-)|(\/)|(\*)|(\()|(\))/
```
```suggestion
    split_pattern = Regexp.union(allowed_chars)
```

EinLama replied:
Didn't know about `Regexp.union`, thanks! That's a nice simplification.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+    # it prevents built-in functions from being used in the formula, which we do not want to allow.
+    allowed_chars = %w[+ - / * ( )] + [" "]
+    split_pattern = /\s|(\+)|(-)|(\/)|(\*)|(\()|(\))/
+    allowed_tokens = /^(cf_\d+|\d+\.?\d*|\.\d+|[#{allowed_chars.join}]+)$/
```
- `allowed_chars` are not needed here, as they can't be present after used for splitting

- `\d+\.?\d*` allows zero digits after decimal point, maybe `\d+\.?\d*|\.\d+` can be replaced with `\d*\.\d+`

- and better to use `\A…\z`

```suggestion
    allowed_tokens = /\A(cf_\d+|\d*\.\d+)\z/
```

EinLama replied:
I think the issue with trailing decimals after the decimal point is caught by Dentaku. I have added a spec for this case.

With regards to the other suggestions:

* Used `\A\z` over `^$`
* With the usage of `Regexp.union`, the allowed chars can be omitted now, yes. In the old regex, I caught them in capture groups so they would have been visible after the split. But since we don't do anything further with the split result, it's better to use your suggestion 👍 Done! 😀

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+
+  # Returns the formula as a string. Will return an empty string if the formula is not set.
+  def formula_string
+    formula.is_a?(Hash) ? formula.fetch("formula", "") : formula.to_s
```
If it is not a Hash, it should probably return `""` or `nil`

toy commented on app/models/project_custom_field.rb:
```diff
@@ -40,12 +40,18 @@ class ProjectCustomField < CustomField
       else
-        where(admin_only: false).where(mappings_with_view_project_attributes_permission(user, project).exists)
+        cfs.where.not(field_format: "calculated_value")
       end
```
Same, I would not hide already created fields

toy commented on spec/lib/open_project/custom_field_format_spec.rb:
```diff
@@ -41,47 +41,55 @@
+                        %w[bool calculated_value date float int link list string text user version]
+      end
+
+      context "with calculated values feature flag disabled" do
```
From experience of removing feature flag, it is nicer when disabled flag is explicit:

```suggestion
      context "with calculated values feature flag disabled", with_flag: { calculated_value_project_attribute: false } do
```

toy commented on spec/lib/open_project/custom_field_format_spec.rb:
```diff
@@ -93,6 +101,15 @@
                               "version", "hierarchy", "empty")
       end
+
+      context "with calculated values feature flag enabled", with_flag: { calculated_value_project_attribute: true } do
```
Nesting inside `custom_field_hierarchies` is not needed - enough to just have one test when nothing is enabled, one when hierarchies and one when calculated values
And it would probably be better to structure tests in `describe ".available_formats"` same as in `describe ".available_for_class_name"` - same idea, less verbose

toy commented on spec/models/custom_field/calculated_value_spec.rb:
```diff
@@ -0,0 +1,118 @@
+    describe "#formula_string" do
+      it "returns an empty string if no formula is set" do
+        subject.formula = nil
+        expect(subject.formula_string).to eq("")
```
I'd expect it to return `nil`

EinLama replied:
I think it's nice to have it return an empty string instead of nil, since you could always call string-chain methods directly on the return value without having to check for nil first. It even says `_string` in its method name, so it should ensure to return a string.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,96 @@
+
+  # Returns a list of custom field IDs used in the formula.
+  # For a formula like `2 + cf_12 + cf_4` it returns `[12, 4]`.
+  def cf_ids_used_in_formula(formula_str = formula_string)
```
As this is supposed to be used only to extract and cache, default argument value is not needed and may be confusing

EinLama replied:
True! This was still a leftover from an earlier implementation that I did not change. The default argument value is indeed confusing. Removed as suggested 👍

toy reviewed (changes requested):
Looks good overall, mostly nitpicks, form is a bit wrong in mockups, I suggest separate table for related fields and not hiding custom fields when flag is off

EinLama committed:
[#64347] remove unnecessary default argument

EinLama committed:
[#64347] is_for_all -> is_required

EinLama committed:
[#64347] use shared examples for #available_formats

EinLama committed:
[#64347] remove unnecessary stub

EinLama committed:
[#64347] remove formula field from regular custom fields

EinLama committed:
[#64347] all.collect.each -> find_each

EinLama committed:
[#64347] name in database cannot be falsy

EinLama committed:
[#64347] use Regexp.union

EinLama committed:
[#64347] simplify formula validation

EinLama committed:
[#64347] use word boundaries in custom field regex

EinLama committed:
[#64347] use Dentaku instance instead of shortcut

EinLama committed:
[#64347] Allow viewing/editing calculated values without FF

We still show these CFs, but we will not allow creating new ones. This
explictly only applies to enabled/disabled custom fields. We are
stricter with the enterprise flag.

EinLama committed:
[#64347] button: primary

EinLama committed:
[#64347] use concern helpers

EinLama committed:
[#64347] move details form to `app/forms`

ulferts commented on app/forms/custom_fields/calculated_values/details_form.rb:
```diff
@@ -0,0 +1,80 @@
+        )
+
+        details_form.submit(name: :submit, label: I18n.t(:button_save), scheme: :primary)
+      end
```
I am not opposed to merging it like that. But we'll have to see what this evolves into. If already have the `CustomFields::DetailsForm` which shares a lot of the fields. 

But let's wait and see what the forms evaluate to when the rest of the formats are migrated as well.

ulferts commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,101 @@
+      # Dentaku will return nil if the formula is invalid.
+      # TODO WP-64348: add support for referenced custom fields by injecting them as variables,
+      #       e.g. calculator.evaluate(formula_string, cf_123: CustomField.find(123).value)
+      errors.add(:formula, :invalid) unless Dentaku::Calculator.new.evaluate(formula_string)
```
I think it is better to rely on Dentaku's AST capabilities for most of the validations. Just because I think they will have worked out more edge cases than we did. So we might as well profit from them - at the cost of binding us more strongly to the gem. 

But we could do the following:

```
ast = Dentaku::Calculator.new.ast(formula_string)

if ast.class.is_a?(Dentaku::AST::Nil) # handles "", " ", "             "
  error.add(:formula, :blank)
  return
end

if ast_contains_unsupported_nodes?(ast) # check for supported operators and that variables follow cf_[id] format
  errors.add(:formula, :invalid_characters)
  return
end

rescue Dentaku::ParseError, [other errors]
  errors.add(:formula, :invalid)
```

That way, the todo's will become reachable I think.

Can also be done in a later PR.

ulferts commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -0,0 +1,101 @@
+    # Returns a list of custom field IDs used in the formula.
+    # For a formula like `2 + cf_12 + cf_4` it returns `[12, 4]`.
+    def cf_ids_used_in_formula(formula_str)
+      formula_str.scan(/\bcf_(\d+)\b/).flatten.map(&:to_i)
```
This could then also be fetched from the AST.

ulferts commented on spec/features/admin/custom_fields/projects/calculated_value_spec.rb:
```diff
@@ -0,0 +1,143 @@
+        expect(page).to have_link("Projects")
+        expect(page).to have_link("Project attributes")
+        expect(page).to have_text("Updated name")
+      end
```
Since we will be working on those custom fields for some time, we should consider investing in creating a spec support page so that we don't duplicate that much code and improve readability.

ulferts reviewed (approved):
The PR looks good @EinLama. 

I made a few remarks but those are targeting future PRs so nothing that would hold us back from merging this one now.
